packages/riverpod/test/devtool_test.dart

```diff
@@ -0,0 +1,218 @@
+    final container = ProviderContainer();
+    final firstProvider = Provider(
+      (_) => 'some string',
+      name: 'firstProvider',
+    );
+    final secondProvider = Provider(
+      (_) => 'some other string',
+      name: 'secondProvider',
+    );
+    container.read(firstProvider);
+    container.read(secondProvider);
+
+    final secondProviderId = container.debugIdForProvider(secondProvider);
+
+    expect(
+      RiverpodBinding.debugInstance.getProvider(
+        container.debugId,
+        secondProviderId,
+      ),
+      predicate(
+        (RiverpodNode node) => node.isEqualTo(
+          RiverpodNode(
+            id: secondProviderId,
+            containerId: container.debugId,
+            type: secondProvider.runtimeType.toString(),
+            mightBeOutdated: false,
+            name: 'secondProvider',
+            state: Result<dynamic>.data('some other string'),
+          ),
+        ),
+      ),
+    );
+  });
+
+  test('should send debug event when provider list changes', () {
+    final spy = spyPostEvent();
+    final container = ProviderContainer();
+    final provider = Provider((_) => 'some string');
+    container.read(provider);
+
+    expect(
+      spy.logs.last,
+      isA<PostEventCall>()
+          .having(
+            (p) => p.eventKind,
+            'eventKind',
+            equals('riverpod:provider_list_changed'),
+          )
+          .having(
+            (p) => p.event,
+            'event',
+            equals({'container_id': container.debugId}),
+          ),
+    );
+  });
+
+  test('should send debug event when container list changes', () {
+    final spy = spyPostEvent();
+    ProviderContainer();
+
+    expect(
+      spy.logs.last,
+      isA<PostEventCall>().having(
+        (p) => p.eventKind,
+        'eventKind',
+        equals('riverpod:container_list_changed'),
+      ),
+    );
+  });
+
+  test('should send debug event on provider listened', () {
+    final spy = spyPostEvent();
+    final container = ProviderContainer();
+    final provider = Provider((_) => 'some string');
+
+    container.listen(provider, (previous, next) {});
+
+    final providerId = container.debugIdForProvider(provider);
+
+    expect(
+      spy.logs.last,
+      isA<PostEventCall>()
+          .having(
+            (p) => p.eventKind,
+            'eventKind',
+            equals('riverpod:provider_changed'),
+          )
+          .having(
+            (p) => p.event,
+            'event',
+            equals({
+              'container_id': container.debugId,
+              'provider_id': providerId,
+            }),
+          ),
+    );
+  });
+}
+
+extension on Map<String, ContainerNode> {
+  bool isEqualTo(Map<String, ContainerNode> other) {
+    return entries.fold(
+      true,
+      (previousValue, entry) {
+        return entry.value.isEqualTo(other[entry.key]!);
+      },
+    );
+  }
+}
+
+extension on ContainerNode {
+  bool isEqualTo(ContainerNode other) {
```

Comment:
we could use `freezed` or `equatable` on the actual classes, but I noticed that `riverpod` package doesn't depend on these packages.